### PR TITLE
Duplex tcp

### LINKED ISSUE #4: TCP connection to simulation is one-way only
Please see https://github.com/ProfHoekstra/bluesky/issues/94

### PR DESCRIPTION
**Pull request by reube**
Resolves #4 - there are some notes on this issue.

### CODE REVIEW

File: bluesky/stack/stack.py [hunk 1]
```diff
@@ -1208,27 +1208,28 @@ def process():
...
                     synerr = not results[0]
                     if synerr:
-                        bs.scr.echo("Syntax error: " + (helptext if len(results) < 2 else ""))
+                        bs.scr.echo("Syntax error: " + (helptext if len(results) < 2 else ""), sender_id)
                     # Maybe there is also an error/info message returned?
                     if len(results) >= 2:
-                        bs.scr.echo(cmd + ":" + results[1])
+                        prefix = "" if results[0] == bs.SIMPLE_ECHO \
+                            else "{}: ".format(cmd)
+                        bs.scr.echo("{}{}".format(prefix, results[1]), sender_id)
```
Comment (reube) [inline bluesky/stack/stack.py:1214]: ^^ We don't always want to prefix the output with the command just entered.

File: bluesky/stack/stack.py [hunk 2]
```diff
@@ -1508,7 +1509,7 @@ def parse(self, argtype, argidx, args):
 def distcalc(lat0, lon0, lat1, lon1):
     try:
         qdr, dist = geo.qdrdist(lat0, lon0, lat1, lon1)
-        return True, "QDR = %.2f deg, Dist = %.3f nm" % (qdr % 360., dist)
+        return bs.SIMPLE_ECHO, "QDR = %.2f deg, Dist = %.3f nm" % (qdr % 360., dist)
     except:
```
Comment (reube) [inline bluesky/stack/stack.py:1500]: ^^ The output when echo'd does not need to repeat the command.  See `stack.process()`

File: bluesky/tools/network.py [hunk 3]
```diff
@@ -143,5 +170,11 @@ def __init__(self):
...
         self.process = fun
 
-    def processData(self, sender_id, data):
-        self.process(bytearray(data).decode(encoding='ascii', errors='ignore').strip())
+    def processData(self, data, sender_id):
+        msg = bytearray(data).decode(encoding='ascii', errors='ignore').strip()
+
+        if msg.startswith(CMD_TCP_CONNS):
+            self.connections[sender_id].sendReply(
+                str(self.numConnections()))
```
Comment (reube) [inline bluesky/tools/network.py:169]: Command TCP_CONNS can be used to debug number of active tcp sessions, and simply returns this information.

File: BlueSky_pygame.py [hunk 4]
```diff
@@ -38,4 +41,4 @@ def MainLoop():
 # Run mainloop if BlueSky_pygame is called directly
 
 if __name__ == '__main__':
-    MainLoop()
+    mainloop()
```
Comment (jacobythwaites) [inline BlueSky_pygame.py]: Prolly a lot of these case-y problems. I like the idea of gently cleaning them out as we go, but may get pushback. I'd use `main_loop` for equivalence with their CamelCasing (camel_casing).

File: BlueSky_qtgl.py [hunk 5]
```diff
@@ -29,52 +34,77 @@ def exception_handler(exc_type, exc_value, exc_traceback):
...
+    # ======================================================================
+    global gui
+    del gui
+    print('BlueSky normal end.')
+
+
+# =============================================================================
+# Start the mainloop (and possible other threads)
+# =============================================================================
+def mainloop():
```
Comment (jacobythwaites) [inline BlueSky_qtgl.py]: main_loop

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,3 @@
+.PHONY: test
```
Comment (jacobythwaites) [inline Makefile:1]: Ah, a makefile :)

File: bluesky/__init__.py [hunk 7]
```diff
@@ -16,3 +16,8 @@
 navdb = _navdb.navdb
 sim   = _sim.sim
 scr   = _scr.scr
+
+SIMPLE_ECHO = 'simple_echo'
+MSG_OK = "ok."
```
Comment (jacobythwaites) [inline bluesky/__init__.py]: Use single or double, I'd not switch between them. Their standard seems to be single which we shd stick with.

File: bluesky/sim/qtgl/mainmanager.py [hunk 8]
```diff
@@ -14,7 +14,7 @@
 # Local imports
 from bluesky import settings
 from .simevents import SimStateEventType, SimQuitEventType, BatchEventType, \
-    BatchEvent, StackTextEvent, SimQuitEvent, SetNodeIdType, \
+    BatchEvent, StackTextEvent, StackTextEventType, SimQuitEvent, SetNodeIdType, \
```
Comment (jacobythwaites) [inline bluesky/sim/qtgl/mainmanager.py:17]: > 80 chars?

File: bluesky/stack/stack.py [hunk 9]
```diff
@@ -1166,7 +1166,7 @@ def process():
                             # No value = None when this is allowed because it is an optional argument
                             if parser.result[0] is None and argisopt[curtype] is False:
                                 synerr = True
-                                bs.scr.echo('No value given for mandatory argument ' + argtypes[curtype])
+                                bs.scr.echo('No value given for mandatory argument ' + argtypes[curtype], sender_id)
```
Comment (jacobythwaites) [inline bluesky/stack/stack.py:1169]: Lot of indent going on here. Not a biggie.

File: bluesky/test/__init__.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
+"""
+Common test functionality.
+- Andrew Farrell, SPARKL.
```
Comment (jacobythwaites) [inline bluesky/test/__init__.py]: Standard heading with copyright per their chosen license for new files

File: bluesky/test/tcp/test_simple.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
+"""
+Tests TCP interface.
```
Comment (jacobythwaites) [inline bluesky/test/tcp/test_simple.py]: Use standard copyright etc preamble for new files per previous

File: bluesky/test/tcp/test_simple.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+BLUESKY = "BlueSky_qtgl.py"
+TCP_HOST = "127.0.0.1"
+TCP_PORT = 8888
+BUFFER_SIZE = 1024
+
+SYNTAX_ERROR = "Syntax error"
+
+
+@pytest.fixture(scope="module")
```
Comment (jacobythwaites) [inline bluesky/test/tcp/test_simple.py:25]: Ah, bliss :)

File: bluesky/tools/network.py [hunk 13]
```diff
@@ -2,35 +2,39 @@
 import time
 import socket
 import threading
-from bluesky import settings
+from bluesky import settings, CMD_TCP_CONNS
+
+SIZEOF_UINT16 = 2
```
Comment (jacobythwaites) [inline bluesky/tools/network.py]: Where is this used?

File: bluesky/traf/route.py [hunk 14]
```diff
@@ -323,7 +323,7 @@ def atwptStack(self, idx, *args):  # args: all arguments of addwpt
                         elif self.wptype[wpidx] == self.dest:
                             txt = txt + "[dest]"
 
-                    return True, txt
+                    return bs.SIMPLE_ECHO, txt
```
Comment (jacobythwaites) [inline bluesky/traf/route.py:326]: Wish I didn't keep hearing bullsh*t every time I see `bs`!

**Review by jacobythwaites — COMMENTED** [review]
Excellent, especially the tests and fledgling makefile. Few minor points only.

File: BlueSky_pygame.py (hunk 4 above)
Comment (reube) [reply BlueSky_pygame.py]: Ack.

File: bluesky/stack/stack.py (hunk 9 above)
Comment (reube) [reply bluesky/stack/stack.py:1169]: That'll be fixed with pepping.

File: bluesky/test/__init__.py (hunk 10 above)
Comment (reube) [reply bluesky/test/__init__.py]: Ack.

File: bluesky/tools/network.py (hunk 13 above)
Comment (reube) [reply bluesky/tools/network.py]: Ah - meant to get rid.  Thx

File: bluesky/sim/qtgl/mainmanager.py (hunk 8 above)
Comment (reube) [reply bluesky/sim/qtgl/mainmanager.py:17]: This is universally problematic at the moment - need to 'pep' the whole source tree.

**Comment by euvl** [comment]
- [x]  LGTM

**Pull request merged**